framework/src/utils/MooseMeshCuttingUtils.C

```diff
@@ -0,0 +1,918 @@
+  return angles;
+}
+
+std::vector<std::pair<Real, unsigned int>>
+vertex_distances(Elem & elem)
+{
+  std::vector<std::pair<Real, unsigned int>> distances;
+  const unsigned int n_vertices = elem.n_vertices();
+
+  for (unsigned int i = 0; i < n_vertices; i++)
+  {
+    Point v1 = (*elem.node_ptr((i + 1) % n_vertices) - *elem.node_ptr(i % n_vertices));
+    distances.push_back(std::make_pair(v1.norm(), i));
+    std::sort(distances.begin(), distances.end());
+  }
+  return distances;
+}
+
+void
+triElemSplitter(ReplicatedMesh & mesh,
+                const dof_id_type elem_id,
+                const unsigned short node_shift,
+                const dof_id_type nid_3,
+                const dof_id_type nid_4,
+                const subdomain_id_type single_elem_side_id,
+                const subdomain_id_type double_elem_side_id,
+                const boundary_id_type new_boundary_id)
+{
+  const auto elem_old = mesh.elem_ptr(elem_id);
+  const dof_id_type nid_0 = elem_old->node_ptr(node_shift % 3)->id();
+  const dof_id_type nid_1 = elem_old->node_ptr((1 + node_shift) % 3)->id();
+  const dof_id_type nid_2 = elem_old->node_ptr((2 + node_shift) % 3)->id();
+
+  const bool m1_side_flag =
+      MooseUtils::absoluteFuzzyEqual((*(mesh.node_ptr(nid_3)) - *(mesh.node_ptr(nid_0)))
+                                         .cross(*(mesh.node_ptr(nid_1)) - *(mesh.node_ptr(nid_0)))
+                                         .norm(),
+                                     0.0);
+  const dof_id_type nid_m1 = m1_side_flag ? nid_3 : nid_4;
+  const dof_id_type nid_m2 = m1_side_flag ? nid_4 : nid_3;
+  // Build boundary information of the mesh
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+  auto bdry_side_list = boundary_info.build_side_list();
+  // Create a list of sidesets involving the element to be split
+  std::vector<std::vector<boundary_id_type>> elem_side_list;
+  elem_side_list.resize(3);
+  for (unsigned int i = 0; i < bdry_side_list.size(); i++)
+  {
+    if (std::get<0>(bdry_side_list[i]) == elem_id)
+    {
+      elem_side_list[(std::get<1>(bdry_side_list[i]) + 3 - node_shift) % 3].push_back(
+          std::get<2>(bdry_side_list[i]));
+    }
+  }
+
+  Elem * elem_Tri3_0 = mesh.add_elem(new Tri3);
+  elem_Tri3_0->set_node(0) = mesh.node_ptr(nid_0);
+  elem_Tri3_0->set_node(1) = mesh.node_ptr(nid_m1);
+  elem_Tri3_0->set_node(2) = mesh.node_ptr(nid_m2);
+  elem_Tri3_0->subdomain_id() = single_elem_side_id;
+  Elem * elem_Tri3_1 = mesh.add_elem(new Tri3);
+  elem_Tri3_1->set_node(0) = mesh.node_ptr(nid_1);
+  elem_Tri3_1->set_node(1) = mesh.node_ptr(nid_m2);
+  elem_Tri3_1->set_node(2) = mesh.node_ptr(nid_m1);
+  elem_Tri3_1->subdomain_id() = double_elem_side_id;
+  Elem * elem_Tri3_2 = mesh.add_elem(new Tri3);
+  elem_Tri3_2->set_node(0) = mesh.node_ptr(nid_2);
+  elem_Tri3_2->set_node(1) = mesh.node_ptr(nid_m2);
+  elem_Tri3_2->set_node(2) = mesh.node_ptr(nid_1);
+  elem_Tri3_2->subdomain_id() = double_elem_side_id;
```

Comment:
It will follow the order of the original TRI3 element.

2D TRI always has a positive volume. Order is more important during extrusion. I think `AdvancedExtruderGenerator` can fix negative element on the fly.